app/src/main/java/org/p2p/wallet/auth/interactor/MetadataInteractor.kt

```diff
@@ -31,30 +36,162 @@ class MetadataInteractor(
         } catch (error: Throwable) {
-            Timber.e(GetOnboardingMetadataFailed.OnboardingMetadataRequestFailure(error))
+            Timber.e(MetadataFailed.OnboardingMetadataRequestFailure(error))
+        }
+    }
+
+    private suspend fun tryToUploadMetadata(metadata: GatewayOnboardingMetadata) {
+        val web3SignUpDetails = signUpDetailsStorage.getLastSignUpUserDetails()?.signUpDetails
+        if (web3SignUpDetails != null) {
+            val userAccount = Account(tokenKeyProvider.keyPair)
+            val userSeedPhrase = seedPhraseProvider.getUserSeedPhrase()
+            tryToUploadMetadata(
+                userAccount = userAccount,
+                mnemonicPhraseWords = userSeedPhrase.seedPhrase,
+                ethereumPublicKey = web3SignUpDetails.ethereumPublicKey,
+                metadata = metadata
+            )
+        } else {
+            Timber.i("User doesn't have any Web3Auth sign up data, skipping metadata fetch")
+        }
+    }
+
+    private suspend fun tryToUploadMetadata(
+        userAccount: Account?,
+        mnemonicPhraseWords: List<String>,
+        ethereumPublicKey: String,
+        metadata: GatewayOnboardingMetadata
+    ) {
+        try {
+            if (userAccount == null) {
+                throw MetadataFailed.MetadataNoAccount()
+            }
+            if (mnemonicPhraseWords.isEmpty()) {
+                throw MetadataFailed.MetadataNoSeedPhrase()
+            }
+            gatewayServiceRepository.updateMetadata(
+                solanaPublicKey = userAccount.publicKey.toBase58Instance(),
+                solanaPrivateKey = userAccount.keypair.toBase58Instance(),
+                userSeedPhrase = mnemonicPhraseWords,
+                ethereumAddress = ethereumPublicKey,
+                metadata = metadata
+            )
+        } catch (cancelled: CancellationException) {
+            Timber.i(cancelled)
+        } catch (validationError: MetadataFailed) {
+            Timber.e(validationError, "Update metadata failed")
+        } catch (error: Throwable) {
+            Timber.e(MetadataFailed.OnboardingMetadataRequestFailure(error))
+        }
+    }
+
+    private suspend fun compareMetadataAndUpdate(serverMetadata: GatewayOnboardingMetadata) {
+        val finalMetadata = secureStorageContract.getObject(
+            SecureStorageContract.Key.KEY_ONBOARDING_METADATA,
+            GatewayOnboardingMetadata::class
+        )?.let { deviceMetadata ->
+            val updatedMetadata = getUpdatedMergedMetadata(deviceMetadata, serverMetadata)
+            if (updatedMetadata != serverMetadata) {
+                tryToUploadMetadata(updatedMetadata)
+            }
+            updatedMetadata
+        } ?: serverMetadata
+        currentMetadata = finalMetadata
+        secureStorageContract.saveObject(SecureStorageContract.Key.KEY_ONBOARDING_METADATA, finalMetadata)
+    }
+
+    private fun getUpdatedMergedMetadata(
```

Comment:
Holy moly

extract this logic to a class asap
you can write tests for checking the update and merge

Reply:
Give me a class name ASAP =)

Reply:
GatewayOnboardingMetadataMerger :) 
with method `merge(local: GatewayOnboardingMetadata, remote: GatewayOnboardingMetadata)`